diff --git a/string/aarch64/memcmp.S b/string/aarch64/memcmp.S
index dac9147..78d9a4f 100644
--- a/string/aarch64/memcmp.S
+++ b/string/aarch64/memcmp.S
@@ -132,3 +132,4 @@ L(byte_loop):
 
 END (__memcmp_aarch64)
 
+#include "symauth.h"
diff --git a/string/aarch64/memrchr.S b/string/aarch64/memrchr.S
index ad42b49..3159626 100644
--- a/string/aarch64/memrchr.S
+++ b/string/aarch64/memrchr.S
@@ -114,3 +114,4 @@ L(nomatch):
 
 END (__memrchr_aarch64)
 
+#include "symauth.h"
diff --git a/string/aarch64/strnlen.S b/string/aarch64/strnlen.S
index 4852edc..d8c6e17 100644
--- a/string/aarch64/strnlen.S
+++ b/string/aarch64/strnlen.S
@@ -108,3 +108,4 @@ L(nomatch):
 
 END (__strnlen_aarch64)
 
+#include "symauth.h"
