diff --git a/osi/src/alarm.cc b/osi/src/alarm.cc
index 1170880d6..c9264e317 100644
--- a/osi/src/alarm.cc
+++ b/osi/src/alarm.cc
@@ -610,7 +610,7 @@ static void alarm_queue_ready(fixed_queue_t* queue, UNUSED_ATTR void* context) {
 }
 
 // Callback function for wake alarms and our posix timer
-static void timer_callback(UNUSED_ATTR void* ptr) {
+static void timer_callback(UNUSED_ATTR sigval_t sigval) {
   semaphore_post(alarm_expired);
 }
 
@@ -678,7 +678,7 @@ static bool timer_create_internal(const clockid_t clock_id, timer_t* timer) {
 
   memset(&sigevent, 0, sizeof(sigevent));
   sigevent.sigev_notify = SIGEV_THREAD;
-  sigevent.sigev_notify_function = (void (*)(union sigval))timer_callback;
+  sigevent.sigev_notify_function = timer_callback;
   sigevent.sigev_notify_attributes = &thread_attr;
   if (timer_create(clock_id, &sigevent, timer) == -1) {
     LOG_ERROR("%s unable to create timer with clock %d: %s", __func__, clock_id,
