diff --git a/libcutils/arch-arm64/android_memset.S b/libcutils/arch-arm64/android_memset.S
index 9a83a6876..f1fe267ac 100644
--- a/libcutils/arch-arm64/android_memset.S
+++ b/libcutils/arch-arm64/android_memset.S
@@ -52,11 +52,18 @@
 #define A_lw		w1
 #define tmp3w		w9
 
+#ifdef ANDROID_EXPERIMENTAL_PAC
+#define AUTH(name, a) .auth name, a
+#else
+#define AUTH(name, a)
+#endif
+
 #define ENTRY(f) \
   .text; \
   .globl f; \
   .align 0; \
   .type f, %function; \
+  AUTH(f, 0x80000000); \
   f: \
   .cfi_startproc \
 
diff --git a/libunwindstack/RegsArm64.cpp b/libunwindstack/RegsArm64.cpp
index 5b7431a57..2a49a55a3 100644
--- a/libunwindstack/RegsArm64.cpp
+++ b/libunwindstack/RegsArm64.cpp
@@ -45,6 +45,12 @@ uint64_t RegsArm64::sp() {
 }
 
 void RegsArm64::set_pc(uint64_t pc) {
+#if defined(__aarch64__)
+  register uint64_t x30 __asm("x30") = pc;
+  // This is XPACLRI.
+  asm("hint 0x7" : "+r"(x30));
+  pc = x30;
+#endif
   regs_[ARM64_REG_PC] = pc;
 }
 
diff --git a/rootdir/init.zygote64_32.rc b/rootdir/init.zygote64_32.rc
index 3e8016867..756c251be 100644
--- a/rootdir/init.zygote64_32.rc
+++ b/rootdir/init.zygote64_32.rc
@@ -5,6 +5,8 @@ service zygote /system/bin/app_process64 -Xzygote /system/bin --zygote --start-s
     group root readproc reserved_disk
     socket zygote stream 660 root system
     socket usap_pool_primary stream 660 root system
+    socket zygote_arm64_v8a stream 660 root system
+    socket usap_pool_arm64_v8a stream 660 root system
     onrestart write /sys/power/state on
     onrestart restart audioserver
     onrestart restart cameraserver
