diff --git a/JniConstants.c b/JniConstants.c
index 08660e0..55481ae 100644
--- a/JniConstants.c
+++ b/JniConstants.c
@@ -123,7 +123,7 @@ static jfieldID FindField(JNIEnv* env, jclass cls,
 static pthread_once_t g_initialized = PTHREAD_ONCE_INIT;
 static JNIEnv* g_init_env;
 
-static void InitializeConstants() {
+static void InitializeConstants(void) {
     // Initialize cached classes.
 #define JCLASS_INITIALIZE(cls, signature, androidOnly)                      \
     CLASS_NAME(cls) = FindClass(g_init_env, signature, androidOnly);
