diff --git a/Android.bp b/Android.bp
index 5783306..aa98f2e 100644
--- a/Android.bp
+++ b/Android.bp
@@ -73,6 +73,9 @@ cc_library_static {
         arm: {
             enabled: true,
         },
+        arm64: {
+            enabled: true,
+        },
     },
     min_sdk_version: "apex_inherit",
 }
diff --git a/src/UnwindRegistersSave.S b/src/UnwindRegistersSave.S
index c7c4959..1999074 100644
--- a/src/UnwindRegistersSave.S
+++ b/src/UnwindRegistersSave.S
@@ -975,3 +975,5 @@ DEFINE_LIBUNWIND_FUNCTION(unw_getcontext)
 #endif /* !defined(__USING_SJLJ_EXCEPTIONS__) */
 
 NO_EXEC_STACK_DIRECTIVE
+
+#include "symauth.h"
