diff --git a/core/binary.mk b/core/binary.mk
index b14d8b82b..d29aed1e5 100644
--- a/core/binary.mk
+++ b/core/binary.mk
@@ -174,6 +174,11 @@ ifneq ($(LOCAL_SDK_VERSION),)
   my_ndk_stl_shared_lib_fullpath :=
   my_ndk_stl_static_lib :=
   my_cpu_variant := $(TARGET_$(LOCAL_2ND_ARCH_VAR_PREFIX)CPU_ABI)
+  ifneq (,$(TARGET_EXPERIMENTAL_PAC))
+    ifeq (arm64-v8a,$(my_cpu_variant))
+      my_cpu_variant := arm64-v9a
+    endif
+  endif
   LOCAL_NDK_STL_VARIANT := $(strip $(LOCAL_NDK_STL_VARIANT))
   ifeq (,$(LOCAL_NDK_STL_VARIANT))
     LOCAL_NDK_STL_VARIANT := system
diff --git a/core/cxx_stl_setup.mk b/core/cxx_stl_setup.mk
index a2abb1a32..3c2ee4020 100644
--- a/core/cxx_stl_setup.mk
+++ b/core/cxx_stl_setup.mk
@@ -82,6 +82,9 @@ ifneq ($(filter $(my_cxx_stl),libc++ libc++_static),)
             ifeq (arm,$($(my_prefix)$(LOCAL_2ND_ARCH_VAR_PREFIX)ARCH))
                 my_static_libraries += libunwind_llvm
                 my_ldflags += -Wl,--exclude-libs,libunwind_llvm.a
+	    else ifeq (arm64,$($(my_prefix)$(LOCAL_2ND_ARCH_VAR_PREFIX)ARCH))
+                my_static_libraries += libunwind_llvm
+                my_ldflags += -Wl,--exclude-libs,libunwind_llvm.a
             else
                 my_static_libraries += libgcc_stripped
                 my_ldflags += -Wl,--exclude-libs,libgcc_stripped.a
@@ -93,6 +96,9 @@ else ifeq ($(my_cxx_stl),ndk)
     ifeq (arm,$($(my_prefix)$(LOCAL_2ND_ARCH_VAR_PREFIX)ARCH))
         my_static_libraries += libunwind_llvm
         my_ldflags += -Wl,--exclude-libs,libunwind_llvm.a
+    else ifeq (arm64,$($(my_prefix)$(LOCAL_2ND_ARCH_VAR_PREFIX)ARCH))
+        my_static_libraries += libunwind_llvm
+        my_ldflags += -Wl,--exclude-libs,libunwind_llvm.a
     else
         my_static_libraries += libgcc_stripped
         my_ldflags += -Wl,--exclude-libs,libgcc_stripped.a
