diff --git a/src/gallium/drivers/freedreno/freedreno_screen.c b/src/gallium/drivers/freedreno/freedreno_screen.c
index 5c0feb9e..3c85a93c 100644
--- a/src/gallium/drivers/freedreno/freedreno_screen.c
+++ b/src/gallium/drivers/freedreno/freedreno_screen.c
@@ -685,7 +685,7 @@ fd_get_compute_param(struct pipe_screen *pscreen, enum pipe_shader_ir ir_type,
 
 static const void *
 fd_get_compiler_options(struct pipe_screen *pscreen,
-		enum pipe_shader_ir ir, unsigned shader)
+		enum pipe_shader_ir ir, enum pipe_shader_type shader)
 {
 	struct fd_screen *screen = fd_screen(pscreen);
 
