diff --git a/cpu_ref/rsCpuIntrinsics_advsimd_3DLUT.S b/cpu_ref/rsCpuIntrinsics_advsimd_3DLUT.S
index edcb0381..475ad3e9 100644
--- a/cpu_ref/rsCpuIntrinsics_advsimd_3DLUT.S
+++ b/cpu_ref/rsCpuIntrinsics_advsimd_3DLUT.S
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#define ENTRY(f) .text; .align 4; .globl f; .type f,#function; f:
+#define ENTRY(f) .text; .align 4; .globl f; .hidden f; .type f,#function; f:
 #define END(f) .size f, .-f;
 
 
diff --git a/cpu_ref/rsCpuIntrinsics_advsimd_Blend.S b/cpu_ref/rsCpuIntrinsics_advsimd_Blend.S
index b4a2b7c4..39cbb72c 100644
--- a/cpu_ref/rsCpuIntrinsics_advsimd_Blend.S
+++ b/cpu_ref/rsCpuIntrinsics_advsimd_Blend.S
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#define ENTRY(f) .text; .align 4; .globl f; .type f,#function; f:
+#define ENTRY(f) .text; .align 4; .globl f; .hidden f; .type f,#function; f:
 #define END(f) .size f, .-f;
 
 #define BLEND_LIST(X) \
diff --git a/cpu_ref/rsCpuIntrinsics_advsimd_Blur.S b/cpu_ref/rsCpuIntrinsics_advsimd_Blur.S
index f3413851..362ee7e2 100644
--- a/cpu_ref/rsCpuIntrinsics_advsimd_Blur.S
+++ b/cpu_ref/rsCpuIntrinsics_advsimd_Blur.S
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#define ENTRY(f) .text; .align 4; .globl f; .type f,#function; f:
+#define ENTRY(f) .text; .align 4; .globl f; .hidden f; .type f,#function; f:
 #define PRIVATE(f) .text; .align 4; .type f,#function; f:
 #define END(f) .size f, .-f;
 
diff --git a/cpu_ref/rsCpuIntrinsics_advsimd_ColorMatrix.S b/cpu_ref/rsCpuIntrinsics_advsimd_ColorMatrix.S
index 55b00298..0288f95a 100644
--- a/cpu_ref/rsCpuIntrinsics_advsimd_ColorMatrix.S
+++ b/cpu_ref/rsCpuIntrinsics_advsimd_ColorMatrix.S
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#define ENTRY(f) .text; .align 4; .globl f; .type f,#function; f:
+#define ENTRY(f) .text; .align 4; .globl f; .hidden f; .type f,#function; f:
 #define END(f) .size f, .-f;
 
 
diff --git a/cpu_ref/rsCpuIntrinsics_advsimd_Convolve.S b/cpu_ref/rsCpuIntrinsics_advsimd_Convolve.S
index 0daa0c5a..c2d08a6d 100644
--- a/cpu_ref/rsCpuIntrinsics_advsimd_Convolve.S
+++ b/cpu_ref/rsCpuIntrinsics_advsimd_Convolve.S
@@ -23,7 +23,7 @@
         x5 = length / 2
 */
 
-#define ENTRY(f) .text; .align 2; .globl f; .type f,#function; f:
+#define ENTRY(f) .text; .align 4; .globl f; .hidden f; .type f,#function; f:
 #define END(f) .size f, .-f;
 
 ENTRY(rsdIntrinsicConvolve3x3_K)
diff --git a/cpu_ref/rsCpuIntrinsics_advsimd_Resize.S b/cpu_ref/rsCpuIntrinsics_advsimd_Resize.S
index 59e735c2..c11b20e1 100644
--- a/cpu_ref/rsCpuIntrinsics_advsimd_Resize.S
+++ b/cpu_ref/rsCpuIntrinsics_advsimd_Resize.S
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#define ENTRY(f) .text; .align 4; .globl f; .type f,#function; f:
+#define ENTRY(f) .text; .align 4; .globl f; .hidden f; .type f,#function; f:
 #define END(f) .size f, .-f;
 
 /* Fixed-point precision after vertical pass -- 16 bit data minus 1 sign and 1
diff --git a/cpu_ref/rsCpuIntrinsics_advsimd_YuvToRGB.S b/cpu_ref/rsCpuIntrinsics_advsimd_YuvToRGB.S
index bb4b7ae3..6f3e3b38 100644
--- a/cpu_ref/rsCpuIntrinsics_advsimd_YuvToRGB.S
+++ b/cpu_ref/rsCpuIntrinsics_advsimd_YuvToRGB.S
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#define ENTRY(f) .text; .align 4; .globl f; .type f,#function; f:
+#define ENTRY(f) .text; .align 4; .globl f; .hidden f; .type f,#function; f:
 #define END(f) .size f, .-f;
 
 /* Perform the actual YuvToRGB conversion in a macro, from register to
