From 72cf3a0c2fd49f46854965bc6aa0aab65f076865 Mon Sep 17 00:00:00 2001
From: Peter Collingbourne <pcc@google.com>
Date: Thu, 11 Jun 2020 11:24:18 -0700
Subject: [PATCH] arm64: Mask out the PAC when entering a signal handler

---
 arch/arm64/kernel/signal.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/kernel/signal.c b/arch/arm64/kernel/signal.c
index 77c02ea5d3b44..5d9bae080e733 100644
--- a/arch/arm64/kernel/signal.c
+++ b/arch/arm64/kernel/signal.c
@@ -728,11 +728,16 @@ static void setup_return(struct pt_regs *regs, struct k_sigaction *ka,
 			 struct rt_sigframe_user_layout *user, int usig)
 {
 	__sigrestore_t sigtramp;
+	unsigned long handler;
 
 	regs->regs[0] = usig;
 	regs->sp = (unsigned long)user->sigframe;
 	regs->regs[29] = (unsigned long)&user->next_frame->fp;
-	regs->pc = (unsigned long)ka->sa.sa_handler;
+
+	handler = (unsigned long)ka->sa.sa_handler;
+	if (system_supports_address_auth())
+		handler &= TASK_SIZE_MAX - 1;
+	regs->pc = handler;
 
 	/*
 	 * Signal delivery is a (wacky) indirect function call in
-- 
2.29.2.729.g45daf8777d-goog

